File: 0040-combination-sum-ii/0040-combination-sum-ii.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    candidates.sort((a,b)=>a-b)
    let ans = []
    solve(0, target, [], ans, candidates)

    return ans
};

var solve = function(i, target, ds, ans, arr){
    if(target === 0){
        ans.push([...ds])
        return
    }

    for(let j=i ; j<arr.length ; j++){
        if(j>i && arr[j] === arr[j-1]) continue
        if(arr[j] > target) break

        ds.push(arr[j])
        solve(j + 1, target - arr[j], ds, ans, arr)
        ds.pop()
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> ans;

    void solve(int i, vector<int>& ds, int target, vector<int>& candidates){
        if(target == 0){
            ans.push_back(ds);
            return;
        }

        for(int j=i ; j<candidates.size() ; j++){
            if(j>i && candidates[j] == candidates[j-1]) continue;
            if(candidates[j] > target) break;

            ds.push_back(candidates[j]);
            solve(j+1, ds, target-candidates[j], candidates);
            ds.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {

        sort(candidates.begin(), candidates.end());
        vector<int> ds;

        solve(0, ds, target, candidates);

        return ans;
    }
};